### PR TITLE
Assert time and audience conditions on SAML responses

### PR DESCRIPTION
**Pull request by mgduk**
SAML responses can (and normally do) supply conditions for when the assertions should be considered valid. `NotBefore` and `NotOnOrAfter` restricts to a time period. `AudienceRestriction` restricts to the intended audience.

The library was previously not validating these at all.

### CODE REVIEW

**Comment by brettkiefer** [comment]
... and the idea is to PR this to clever/saml2 as well?

**Comment by mgduk** [comment]
@brettkiefer yeah — sorry, should have made that clear. Approval to contribute to the project is pending with the legal team, so this PR on our fork is just to facilitate code review before I can open the PR back to clever/saml2.

File: package.json [hunk 1]
```diff
@@ -30,6 +30,7 @@
   "dependencies": {
     "async": "^2.5.0",
     "debug": "^2.6.0",
+    "moment": "^2.18.1",
```
Comment (brettkiefer) [inline package.json]: Including moment for those two calls seems a little bit much -- is there a good reason?

Comment (mgduk) [reply package.json]: Agreed. I've removed that and force pushed.

File: lib/saml2.coffee [hunk 2]
```diff
@@ -595,6 +597,32 @@ module.exports.ServiceProvider =
...
+                    if condition == 'notbefore' and Date.parse(attribute.value) > Date.now()
+                      return cb_wf new SAMLError('SAML Response is not yet valid', {NotBefore: attribute.value})
+                    if condition == 'notonorafter' and Date.parse(attribute.value) <= Date.now()
+                      return cb_wf new SAMLError('SAML Response is no longer valid', {NotOnOrAfter: attribute.value})
+
+                audience_restriction = conditions.getElementsByTagNameNS(XMLNS.SAML, 'AudienceRestriction')[0]
+                audiences = audience_restriction?.getElementsByTagNameNS(XMLNS.SAML, 'Audience')
+                if audiences?.length > 0
+                  validAudience = false
+                  for audience in audiences
```
Comment (brettkiefer) [inline lib/saml2.coffee]: It looks like they're already including underscore .. this feels a lot like an underscore _.find -- `validAudience = _.find(ye predicate)` it'd simplify the code a little bit for the reader and modifier, and I think have all of the same performance characteristics. Would that be idiomatic within this codebase? Minor thing, feel free to brush it off if there are good reasons to do it this way.

Comment (mgduk) [reply lib/saml2.coffee]: Thanks @brettkiefer — it is better with `_.find`. I've pushed a fixup for when you're ready to re-review!

File: lib/saml2.coffee [hunk 3]
```diff
@@ -609,18 +609,15 @@ module.exports.ServiceProvider =
                       return cb_wf new SAMLError('SAML Response is no longer valid', {NotOnOrAfter: attribute.value})
 
                 audience_restriction = conditions.getElementsByTagNameNS(XMLNS.SAML, 'AudienceRestriction')[0]
-                audiences = audience_restriction?.getElementsByTagNameNS(XMLNS.SAML, 'Audience')
+                audiences =  audience_restriction?.getElementsByTagNameNS(XMLNS.SAML, 'Audience')
```
Comment (brettkiefer) [inline lib/saml2.coffee]: Looks like ya picked up an extra space.

Comment (mgduk) [reply lib/saml2.coffee]: Now with less space.

**Review by brettkiefer — APPROVED** [review]

**Comment by mgduk** [comment]
@brettkiefer I've added a commit to allow for clock skew tolerance between servers — see commit message for details: https://github.com/trello/saml2/pull/1/commits/eefbe39f4bdbb7eff896b465b5271e5b82bd39d3

Would you kindly review that additional commit, then I can open the PR with Clever.

File: lib/saml2.coffee [hunk 4]
```diff
@@ -501,12 +501,13 @@ module.exports.ServiceProvider =
       {@entity_id, @private_key, @certificate, @assert_endpoint, @alt_private_keys, @alt_certs} = options
 
       options.audience ?= @entity_id
+      options.notbefore_skew ?= 1
```
Comment (brettkiefer) [inline lib/saml2.coffee]: A default of 1 instead of 0 ?

Comment (brettkiefer) [reply lib/saml2.coffee]: Oh, commit comment. OK.

**Review by brettkiefer — APPROVED** [review]

**Comment by mgduk** [comment]
Closed as this was just for internal review purposes. Continued in a PR to merge back up to the source repo: https://github.com/Clever/saml2/pull/116

**Pull request merged**